skimage/morphology/_watershed.pyx

```diff
@@ -10,87 +10,115 @@ All rights reserved.
+        The raveled coordinates of the initial markers (aka seeds) for the
+        watershed. NOTE: these should *all* point to nonzero entries in the
+        output, or the algorithm will never terminate and blow up your memory!
+    structure : array of int
+        A list of coordinate offsets to compute the raveled coordinates of each
+        neighbor from the raveled coordinates of the current pixel.
+    mask : array of int
+        An array of the same shape as `image` where each pixel contains a
+        nonzero value if it is to be considered for flooding with watershed,
+        zero otherwise. NOTE: it is *essential* that the border pixels (those
+        with neighbors falling outside the volume) are all set to zero, or
+        segfaults could occur.
+    strides : array of int
+        An array representing the number of steps to move along each dimension.
+        This is used in computing the Euclidean distance between raveled
+        indices.
+    compactness : float
+        A value >0 implements the compact watershed algorithm (see .py file).
+    output : array of int
+        The output array, which must already contain nonzero entries at all the
+        seed locations.
     """
     cdef Heapitem elem
     cdef Heapitem new_elem
     cdef Py_ssize_t nneighbors = structure.shape[0]
     cdef Py_ssize_t i = 0
+    cdef Py_ssize_t age = 1
     cdef Py_ssize_t index = 0
-    cdef Py_ssize_t old_index = 0
-    cdef Py_ssize_t max_index = image.shape[0]
 
     cdef Heap *hp = <Heap *> heap_from_numpy2()
 
-    for i in range(pq.shape[0]):
-        elem.value = pq[i, 0]
-        elem.age = pq[i, 1]
-        elem.index = pq[i, 2]
+    for i in range(marker_locations.shape[0]):
+        index = marker_locations[i]
+        elem.value = image[index]
+        elem.age = 0
+        elem.index = index
+        elem.source = index
         heappush(hp, &elem)
 
     while hp.items > 0:
-        #
-        # Pop off an item to work on
-        #
         heappop(hp, &elem)
-        ####################################################
-        # loop through each of the structuring elements
-        #
-        old_index = elem.index
+        if output[elem.index] and elem.index != elem.source:
+            # non-marker, already visited from another neighbor
+            continue
+        output[elem.index] = output[elem.source]
         for i in range(nneighbors):
             # get the flattened address of the neighbor
-            index = structure[i, 0] + old_index
-            if index < 0 or index >= max_index or output[index] or \
-                    not mask[index]:
+            index = structure[i] + elem.index
+            if output[index] or not mask[index]:
+                # previously visited, masked, or border pixel
                 continue
 
             age += 1
             new_elem.value = image[index]
+            if compactness > 0:
+                new_elem.value += (compactness *
```

Comment:
`new_elem.value` is an int32 while the right hand side of the equation is a float. Does this result in artifacts since you loose precision?